js/ui/shellDBus.js

```diff
@@ -592,26 +612,50 @@ const AppLauncher = new Lang.Class({
     },
 
     LaunchAsync: function(params, invocation) {
-        if (name == 'eos-app-store') {
+        let [appName, timestamp] = params;
+
+        if (appName == 'eos-app-store') {
             Main.appStore.show(timestamp, true);
+            Main.appStore.appLaunched = true;
             return;
         }
 
-        let [appName, timestamp] = params;
-        if (!appName.endsWith('.desktop'))
-            appName += '.desktop';
+        let activationContext = activationContextForAppName(appName, this._appSys);
 
-        let app = this._appSys.lookup_app(appName);
-        if (!app) {
+        if (!activationContext) {
             invocation.return_error_literal(Gio.IOErrorEnum,
                                             Gio.IOErrorEnum.NOT_FOUND,
                                             'Unable to launch app ' + appName + ': Not installed');
             return;
         }
 
-        let activationContext = new AppActivation.AppActivationContext(app);
         activationContext.activate(null, timestamp);
+    },
 
+    LaunchViaDBusCallAsync: function(params, invocation) {
+        let [appName, busName, path, interfaceName, method, args] = params;
+        let activationContext = activationContextForAppName(appName, this._appSys);
+
+        if (!activationContext) {
+            invocation.return_error_literal(Gio.IOErrorEnum,
+                                            Gio.IOErrorEnum.NOT_FOUND,
+                                            'Unable to launch app ' + appName + ': Not installed');
+            return;
+        }
+
+        activationContext.activateViaDBusCall(busName, path, interfaceName, method, args, function(error, result) {
+            if (error) {
+                logError(error);
+                invocation.return_error_literal(Gio.IOErrorEnum,
```

Comment:
I don't think so in the sense that we need to convert the error into a GIOError (and should probably wrap it too.